## FIFA 2021 DATA ANALYSIS 
#### Dataset: The data was gotten from kaggle and can be found [Here]('https://www.kaggle.com/datasets/yagunnersya/fifa-21-messy-raw-dataset-for-cleaning-exploring')

> *The dataset consist of various players name including their nationality, club, age, height, assist, e.t.c.*

## Importing Necessary Libraries


In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
import warnings
%matplotlib inline 

### Loading the dataset

In [2]:
data = pd.read_csv('C:/Users/Taiwo_Victoria/Desktop/DATA ANALYSIS PROJECTS/FIFA 21 messy_raw dataset for cleaning exploring/dataset/fifa21_raw_data_v2.csv',low_memory = False )

## Data preparation 

In [3]:
data.head()

ID               Name                      LongName  \
0  158023           L. Messi                  Lionel Messi   
1   20801  Cristiano Ronaldo  C. Ronaldo dos Santos Aveiro   
2  200389           J. Oblak                     Jan Oblak   
3  192985       K. De Bruyne               Kevin De Bruyne   
4  190871          Neymar Jr    Neymar da Silva Santos Jr.   

                                           photoUrl  \
0  https://cdn.sofifa.com/players/158/023/21_60.png   
1  https://cdn.sofifa.com/players/020/801/21_60.png   
2  https://cdn.sofifa.com/players/200/389/21_60.png   
3  https://cdn.sofifa.com/players/192/985/21_60.png   
4  https://cdn.sofifa.com/players/190/871/21_60.png   

                                           playerUrl Nationality  Age  ↓OVA  \
0  http://sofifa.com/player/158023/lionel-messi/2...   Argentina   33    93   
1  http://sofifa.com/player/20801/c-ronaldo-dos-s...    Portugal   35    92   
2  http://sofifa.com/player/200389/jan-oblak/210006/    Slovenia   27    91   
3  http://sofifa.com/player/192985/kevin-de-bruyn...     Belgium   29    91   
4  http://sofifa.com/player/190871/neymar-da-silv...      Brazil   28    91   

   POT                         Club  ...     A/W     D/W   IR PAC SHO  PAS  \
0   93         \n\n\n\nFC Barcelona  ...  Medium     Low  5 ★  85  92   91   
1   92             \n\n\n\nJuventus  ...    High     Low  5 ★  89  93   81   
2   93      \n\n\n\nAtlético Madrid  ...  Medium  Medium  3 ★  87  92   78   
3   91      \n\n\n\nManchester City  ...    High    High  4 ★  76  86   93   
4   91  \n\n\n\nParis Saint-Germain  ...    High  Medium  5 ★  91  85   86   

  DRI DEF PHY Hits  
0  95  38  65  771  
1  89  35  77  562  
2  90  52  90  150  
3  88  64  78  207  
4  94  36  59  595  

[5 rows x 77 columns]

In [4]:
data.shape

(18979, 77)

> There are 77 columns in the dataset. Since all the columns will not be considered, some of the columns will be extracted out.

In [5]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18979 entries, 0 to 18978
Data columns (total 77 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   ID                18979 non-null  int64 
 1   Name              18979 non-null  object
 2   LongName          18979 non-null  object
 3   photoUrl          18979 non-null  object
 4   playerUrl         18979 non-null  object
 5   Nationality       18979 non-null  object
 6   Age               18979 non-null  int64 
 7   ↓OVA              18979 non-null  int64 
 8   POT               18979 non-null  int64 
 9   Club              18979 non-null  object
 10  Contract          18979 non-null  object
 11  Positions         18979 non-null  object
 12  Height            18979 non-null  object
 13  Weight            18979 non-null  object
 14  Preferred Foot    18979 non-null  object
 15  BOV               18979 non-null  int64 
 16  Best Position     18979 non-null  object
 17  Joined      

From this analysis, there are some questions that will be answered from the dataset.

Questions:

*1.* Who is the oldest fifa21 player?

*2.* Who are the 10 most paid players? 

*3.* Which players have been playing at a club for more than 10 years.

*4.* Which players are highly valuable but still underpaid(on low wages).

*5.* What is the highest player loan period?

*6.* Who is the tallest player?

*7.* Which player as the longest conntract with his club?

*8.* Who are the best players based on their positioning?

*9.* How many africans players are playing in Europe?

*10.* Who is the heaviest player?



In [6]:
df = data[['LongName', 'Nationality', 'Age', '↓OVA', 'Club', 'Joined', 'Contract', 'Height', 'Weight', 'Preferred Foot', 'Best Position','Value', 'Loan Date End', 'Wage']]

In [7]:
df.shape

(18979, 14)

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18979 entries, 0 to 18978
Data columns (total 14 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   LongName        18979 non-null  object
 1   Nationality     18979 non-null  object
 2   Age             18979 non-null  int64 
 3   ↓OVA            18979 non-null  int64 
 4   Club            18979 non-null  object
 5   Joined          18979 non-null  object
 6   Contract        18979 non-null  object
 7   Height          18979 non-null  object
 8   Weight          18979 non-null  object
 9   Preferred Foot  18979 non-null  object
 10  Best Position   18979 non-null  object
 11  Value           18979 non-null  object
 12  Loan Date End   1013 non-null   object
 13  Wage            18979 non-null  object
dtypes: int64(2), object(12)
memory usage: 2.0+ MB


## Data Cleaning

In [9]:
df.isnull().sum()

LongName              0
Nationality           0
Age                   0
↓OVA                  0
Club                  0
Joined                0
Contract              0
Height                0
Weight                0
Preferred Foot        0
Best Position         0
Value                 0
Loan Date End     17966
Wage                  0
dtype: int64

#### Longname

In [10]:
# checking for null values
df['LongName'].info()

<class 'pandas.core.series.Series'>
RangeIndex: 18979 entries, 0 to 18978
Series name: LongName
Non-Null Count  Dtype 
--------------  ----- 
18979 non-null  object
dtypes: object(1)
memory usage: 148.4+ KB


In [11]:
#checking for all unique values in the column
df['LongName'].nunique()

18852

In [12]:
#printing out all repeated values in the row
df[df['LongName'].duplicated()]

LongName          Nationality  Age  ↓OVA  \
1239    Lisandro López            Argentina   30    76   
2511        Ben Davies              England   24    73   
2517    Gonzalo Castro              Germany   33    73   
2957     Claudio Bravo            Argentina   23    73   
4275   Ladislav Krejčí       Czech Republic   21    71   
...                ...                  ...  ...   ...   
18707        Peng Wang             China PR   22    51   
18765      Alfie Jones              England   19    50   
18782       Joe Wright              England   19    50   
18822       Mark Byrne  Republic of Ireland   19    50   
18865        Wei Zhang             China PR   20    49   

                                 Club        Joined     Contract Height  \
1239             \n\n\n\nBoca Juniors   Jan 1, 2020  2020 ~ 2023  188cm   
2511        \n\n\n\nPreston North End   Jan 1, 2013  2013 ~ 2021  185cm   
2517            \n\n\n\nVfB Stuttgart   Jul 1, 2018  2018 ~ 2021  172cm   
2957   \n\n\n\nClub Atlético Banfield   Jan 4, 2016  2016 ~ 2021  170cm   
4275             \n\n\n\nSparta Praha   Jul 2, 2019  2019 ~ 2023  190cm   
...                               ...           ...          ...    ...   
18707        \n\n\n\nGuangzhou R&F FC  Aug 10, 2020  2020 ~ 2023  176cm   
18765            \n\n\n\nCrawley Town   Aug 8, 2019  2019 ~ 2021  186cm   
18782                \n\n\n\nMillwall   Jul 1, 2019  2019 ~ 2024  185cm   
18822            \n\n\n\nSligo Rovers  Feb 13, 2020  2020 ~ 2020  178cm   
18865  \n\n\n\nHebei China Fortune FC   Jan 1, 2019  2019 ~ 2024  186cm   

      Weight Preferred Foot Best Position  Value Loan Date End  Wage  
1239    80kg          Right            CB    €6M           NaN  €16K  
2511    74kg           Left            CB  €4.4M           NaN  €18K  
2517    72kg          Right            CM  €1.9M           NaN  €20K  
2957    69kg           Left           LWB  €6.5M           NaN  €10K  
4275    86kg           Left            CB  €3.6M           NaN  €500  
...      ...            ...           ...    ...           ...   ...  
18707   69kg          Right           CAM  €130K           NaN   €2K  
18765   73kg          Right            GK  €110K           NaN  €500  
18782   82kg          Right            GK  €100K           NaN  €650  
18822   74kg          Right           CAM  €110K           NaN  €500  
18865   74kg          Right           CDM  €100K           NaN   €1K  

[127 rows x 14 columns]

In [13]:
#deleting all duplicated values while keeping the original value
df = df.drop_duplicates(subset = ['LongName'])

In [14]:
df['LongName'].head()

0                    Lionel Messi
1    C. Ronaldo dos Santos Aveiro
2                       Jan Oblak
3                 Kevin De Bruyne
4      Neymar da Silva Santos Jr.
Name: LongName, dtype: object

In [15]:
#verifying that there is no duplicated values in the column.
df[df['LongName'].duplicated()]

Empty DataFrame
Columns: [LongName, Nationality, Age, ↓OVA, Club, Joined, Contract, Height, Weight, Preferred Foot, Best Position, Value, Loan Date End, Wage]
Index: []

#### Nationality

In [16]:
df['Nationality'].info()

<class 'pandas.core.series.Series'>
Int64Index: 18852 entries, 0 to 18978
Series name: Nationality
Non-Null Count  Dtype 
--------------  ----- 
18852 non-null  object
dtypes: object(1)
memory usage: 294.6+ KB


In [17]:
df['Nationality'].nunique()

164

In [18]:
df['Nationality'].head()

0    Argentina
1     Portugal
2     Slovenia
3      Belgium
4       Brazil
Name: Nationality, dtype: object

In [19]:
df['Nationality'].shape

(18852,)

In [20]:
df['Nationality'].isnull().sum()

0

*It shows that all the 18852 players under consideration has their own nationality known.*

In [21]:
df.head()

LongName Nationality  Age  ↓OVA  \
0                  Lionel Messi   Argentina   33    93   
1  C. Ronaldo dos Santos Aveiro    Portugal   35    92   
2                     Jan Oblak    Slovenia   27    91   
3               Kevin De Bruyne     Belgium   29    91   
4    Neymar da Silva Santos Jr.      Brazil   28    91   

                          Club        Joined     Contract Height Weight  \
0         \n\n\n\nFC Barcelona   Jul 1, 2004  2004 ~ 2021  170cm   72kg   
1             \n\n\n\nJuventus  Jul 10, 2018  2018 ~ 2022  187cm   83kg   
2      \n\n\n\nAtlético Madrid  Jul 16, 2014  2014 ~ 2023  188cm   87kg   
3      \n\n\n\nManchester City  Aug 30, 2015  2015 ~ 2023  181cm   70kg   
4  \n\n\n\nParis Saint-Germain   Aug 3, 2017  2017 ~ 2022  175cm   68kg   

  Preferred Foot Best Position    Value Loan Date End   Wage  
0           Left            RW  €103.5M           NaN  €560K  
1          Right            ST     €63M           NaN  €220K  
2          Right            GK    €120M           NaN  €125K  
3          Right           CAM    €129M           NaN  €370K  
4          Right            LW    €132M           NaN  €270K

#### Club
> There is a need to remove all the \n in front of the club name for this column.

In [22]:
df['Club'] = df['Club'].map(lambda x: x.lstrip('\n'))

In [23]:
df['Club'].head()

0           FC Barcelona
1               Juventus
2        Atlético Madrid
3        Manchester City
4    Paris Saint-Germain
Name: Club, dtype: object

#### Joined

In [24]:
df['Joined'].info()

<class 'pandas.core.series.Series'>
Int64Index: 18852 entries, 0 to 18978
Series name: Joined
Non-Null Count  Dtype 
--------------  ----- 
18852 non-null  object
dtypes: object(1)
memory usage: 294.6+ KB


This column refers to the date that the players joined the football club they are playing for. 

The column datatype needs to be change to date instead of object.

In [27]:
from datetime import datetime

In [29]:
df['Joined'] = pd.to_datetime(df['Joined'])

In [31]:
df['Joined'].head()

0   2004-07-01
1   2018-07-10
2   2014-07-16
3   2015-08-30
4   2017-08-03
Name: Joined, dtype: datetime64[ns]

#### Contract

In [32]:
df['Contract'].head()

0    2004 ~ 2021
1    2018 ~ 2022
2    2014 ~ 2023
3    2015 ~ 2023
4    2017 ~ 2022
Name: Contract, dtype: object

Considering the format of the *Contract* column, the column will be divided into two column to short the start and the end of the contract then the difference will be calculated to get the exact contract per year. 

In [34]:
# dividing the column into to with a delimeter ( ~ )
df[['Start','Last']] = df.Contract.str.split(" ~ ",expand=True)

In [41]:
df['Start']

0        2004
1        2018
2        2014
3        2015
4        2017
         ... 
18974    2018
18975    2020
18976    2019
18977    2020
18978    2019
Name: Start, Length: 18852, dtype: object

In [38]:
df['Last'].head()

0    2021
1    2022
2    2023
3    2023
4    2022
Name: Last, dtype: object

In [43]:
df['Contract'].unique()

array(['2004 ~ 2021', '2018 ~ 2022', '2014 ~ 2023', '2015 ~ 2023',
       '2017 ~ 2022', '2017 ~ 2023', '2018 ~ 2024', '2014 ~ 2022',
       '2018 ~ 2023', '2016 ~ 2023', '2013 ~ 2023', '2011 ~ 2023',
       '2009 ~ 2022', '2005 ~ 2021', '2011 ~ 2021', '2015 ~ 2022',
       '2017 ~ 2024', '2010 ~ 2024', '2012 ~ 2021', '2019 ~ 2024',
       '2015 ~ 2024', '2017 ~ 2025', '2020 ~ 2025', '2019 ~ 2023',
       '2008 ~ 2023', '2015 ~ 2021', '2020 ~ 2022', '2012 ~ 2022',
       '2016 ~ 2025', '2013 ~ 2022', '2011 ~ 2022', '2012 ~ 2024',
       '2016 ~ 2021', '2012 ~ 2023', '2008 ~ 2022', '2019 ~ 2022',
       '2017 ~ 2021', '2013 ~ 2024', '2020 ~ 2024', '2010 ~ 2022',
       '2020 ~ 2021', '2011 ~ 2024', '2020 ~ 2023', '2014 ~ 2024',
       '2013 ~ 2026', '2016 ~ 2022', '2010 ~ 2021', '2013 ~ 2021',
       '2019 ~ 2025', '2018 ~ 2025', '2016 ~ 2024', '2018 ~ 2021',
       '2009 ~ 2024', '2007 ~ 2022', 'Jun 30, 2021 On Loan',
       '2009 ~ 2021', '2019 ~ 2021', '2019 ~ 2026', 'Free', '2012 ~ 

In [48]:
df['Start'].unique()

array(['2004', '2018', '2014', '2015', '2017', '2016', '2013', '2011',
       '2009', '2005', '2010', '2012', '2019', '2020', '2008', '2007',
       'Jun 30, 2021 On Loan', 'Free', '2002', 'Dec 31, 2020 On Loan',
       'Jan 30, 2021 On Loan', 'Jun 30, 2022 On Loan',
       'May 31, 2021 On Loan', '2006', 'Jul 5, 2021 On Loan',
       'Dec 31, 2021 On Loan', 'Jul 1, 2021 On Loan',
       'Jan 1, 2021 On Loan', 'Aug 31, 2021 On Loan', '2003',
       'Jan 31, 2021 On Loan', 'Dec 30, 2021 On Loan',
       'Jun 23, 2021 On Loan', 'Jan 3, 2021 On Loan',
       'Nov 27, 2021 On Loan', 'Jan 17, 2021 On Loan',
       'Jun 30, 2023 On Loan', '1998', 'Jul 31, 2021 On Loan',
       'Nov 22, 2020 On Loan', 'May 31, 2022 On Loan',
       'Dec 30, 2020 On Loan', 'Jan 4, 2021 On Loan',
       'Nov 30, 2020 On Loan', 'Aug 1, 2021 On Loan'], dtype=object)

In [45]:
df['Last'].unique()

array(['2021', '2022', '2023', '2024', '2025', '2026', None, '2028',
       '2020', '2027'], dtype=object)

In [46]:
# converting the *Start* and *Last* column to int

df['Last'] = pd.to_numeric(df['Last'])

In [ ]:
print